### ISSUE TITLE
Render JSON String Diffs as JSON

### ISSUE THREAD

**Issue by jcjones** [hashicorp/nomad#18103]
### Proposal
In Terraform `plan` and `apply` diff output, when [a string evaluated looks like JSON](https://github.com/hashicorp/terraform/blob/4b2b34929de1f93f02804fddf2432984c62d1514/internal/command/jsonformat/computed/renderers/string.go#L32-L40), the [renderer calls a function `renderStringDiffAsJson` which pretty-prints the output](https://github.com/hashicorp/terraform/blob/4b2b34929de1f93f02804fddf2432984c62d1514/internal/command/jsonformat/computed/renderers/primitive.go#L189).

It would be lovely to have available to Nomad the same capability for understanding diffs of pre-rendered JSON blocks, for when rendering via `jsonencode` is impractical. This would serve to reduce human errors when evaluating the correctness of `plan` outputs.

### Use-cases
Job specifications which render out JSON can be difficult to evaluate in a `plan`. Take for example this change, which changes a setting `maxConnectionAge` from `60s` to `30s` in an inline JSON template:

```
     +/- Job: "boulder-nonce"
     +/- Task Group: "nonce" (1 canary, 1 ignore)
       +/- Task: "boulder-nonce" (forces create/destroy update)
         +/- Template {
               ChangeMode:    "restart"
               ChangeSignal:  ""
               DestPath:      "${NOMAD_TASK_DIR}/nonce.json"
           +/- EmbeddedTmpl:  "{\n  \"NonceService\": {\n    \"maxUsed\": {{ key \"service/nonce/maxUsed\" | parseInt }},\n    \"noncePrefix\": \"{{ slice (env \"NOMAD_ADDR_grpc\" | sha256Hex) 0 4 | toUpper }}\",\n    \"debugAddr\": \"{{ env \"NOMAD_ADDR_debug\" }}\",\n    \"grpc\": {\n      \"address\": \"{{ env \"NOMAD_ADDR_grpc\" }}\",\n      \"maxConnectionAge\": \"60s\",\n      \"services\": {\n        \"nonce.NonceService\": {\n          \"clientNames\": {{ key \"service/nonce/clientNames\" }}\n        },\n        \"grpc.health.v1.Health\": {\n          \"clientNames\": [\n            \"ifn-worker.{{ env \"meta.le_env_service\" }}\"\n          ]\n        }\n      }\n    },\n    \"tls\": {\n      \"caCertFile\": \"secrets/nonce/ca-cert.pem\",\n      \"certFile\": \"secrets/nonce/cert.pem\",\n      \"keyFile\": \"secrets/nonce/key.pem\"\n    },\n    \"syslog\": {{ keyOrDefault \"service/nonce/syslog\" \"{}\" }}\n  }\n}\n" => "{\n  \"NonceService\": {\n    \"maxUsed\": {{ key \"service/nonce/maxUsed\" | parseInt }},\n    \"noncePrefix\": \"{{ slice (env \"NOMAD_ADDR_grpc\" | sha256Hex) 0 4 | toUpper }}\",\n    \"debugAddr\": \"{{ env \"NOMAD_ADDR_debug\" }}\",\n    \"grpc\": {\n      \"address\": \"{{ env \"NOMAD_ADDR_grpc\" }}\",\n      \"maxConnectionAge\": \"30s\",\n      \"services\": {\n        \"nonce.NonceService\": {\n          \"clientNames\": {{ key \"service/nonce/clientNames\" }}\n        },\n        \"grpc.health.v1.Health\": {\n          \"clientNames\": [\n            \"ifn-worker.{{ env \"meta.le_env_service\" }}\"\n          ]\n        }\n      }\n    },\n    \"tls\": {\n      \"caCertFile\": \"secrets/nonce/ca-cert.pem\",\n      \"certFile\": \"secrets/nonce/cert.pem\",\n      \"keyFile\": \"secrets/nonce/key.pem\"\n    },\n    \"syslog\": {{ keyOrDefault \"service/nonce/syslog\" \"{}\" }}\n  }\n}\n"
               Envvars:       "false"
               ErrMissingKey: "false"
               LeftDelim:     "{{"
               Perms:         "0644"
               RightDelim:    "}}"
               SourcePath:    ""
               Splay:         "30000000000"
               VaultGrace:    "0"
             }
           Task: "generate-shutdown-helper"

```

### Attempted Solutions

It's possible to externally decode the `A => B` syntax into a diff, though complicated by embedded consul template statements. By comparison, Terraform's rendering makes such a diff immediately human-readable.

An optimal solution is to not inline JSON, but rather use native HCL structures and the `jsonencode` tool. However, that can be difficult when still relying on runtime lookups or consul templates. So for codebases that are still migrating, JSON rendering would reduce human errors while validating `plan` output.

**Comment by tgross** [comment]
I had no idea Terraform did that, very clever. :grinning: I'll get this into roadmapping, but it also looks like the sort of thing that someone from the community might want to try to pick up so I'll mark it appropriately.

**Comment by PavloGl** [comment]
Hi @tgross ,
Is issue still valid? If yes, I would like to contribute 

Thanks

**Comment by tgross** [comment]
Yes, @PavloGl! You'll find the area of code of interest to be in [`nomad/structs/diff.go`](https://github.com/hashicorp/nomad/blob/main/nomad/structs/diff.go) (with tests in `diff_test.go` that you'll want to add to). Thanks!

**Comment by PavloGl** [comment]
Thanks for heads up!

**Comment by vipulbhasin23** [comment]
Hi @tgross — I'd like to pick this up, following on from #28298. Pretty-printing JSON in fieldDiff would make embedded template diffs even more readable, and both CLI and UI get it from one shared fix. Will add tests to diff_test.go.

**Comment by vipulbhasin23** [comment]
Digging into this a bit more before writing code — wanted to flag a scope question.

The cleanest fix is a single change in `fieldDiff()` (`nomad/structs/diff.go`): if `old`/`new` both parse as valid JSON, pretty-print them (indent) before building the `FieldDiff`. Since both the CLI (`command/job_plan.go`) and the UI (`ui/app/components/job-diff-template.js`) just read `FieldDiff.Old`/`.New`, this is a shared fix rather than duplicating JSON-detection in both places.

One wrinkle: the CLI's `formatFieldDiff` currently uses `%q` to render `Old`/`New` (`"%q => %q"`), which escapes embedded newlines as literal `\n` — so a pretty-printed multi-line JSON value would print as one long line with visible `\n`s, which is worse than today. The line-by-line diff work from #23603 already solves this, but only for the `EmbeddedTmpl` field specifically: it's gated by `diff.Name == "EmbeddedTmpl"` in both the CLI (uses `cmp.Diff` instead of `%q`) and the UI (uses `diffLines`).

So there are two ways to scope this:

1. **Keep the JSON pretty-printing scoped to `EmbeddedTmpl`** — smallest change, directly fixes the template-diff use case from the original report, reuses the existing multi-line-safe rendering path as-is.
2. **Generalize**: pretty-print JSON for any field, and change the CLI/UI gate from "field name is EmbeddedTmpl" to "does this value contain a newline" — matches the issue title more literally (any JSON-valued field, not just templates), but touches more surface (the rendering gate in `command/job_plan.go` and the UI condition, not just `diff.go`).

Leaning towards (1) since it directly addresses the motivating example and doesn't risk garbled output for other fields, but wanted to check if you had a preference before I start. Happy to go with whichever.

**Comment by tgross** [comment]
Sorry for the delayed reply... I've been a bit swamped.

> The line-by-line diff work from #23603 already solves this, but only for the `EmbeddedTmpl` field specifically: it's gated by `diff.Name == "EmbeddedTmpl"` in both the CLI (uses `cmp.Diff` instead of `%q`) and the UI (uses `diffLines`).

Where else would there be JSON? I think this can get closed by #23603, can't it?

**Comment by vipulbhasin23** [comment]
Checked `diff.go` directly — `EmbeddedTmpl` isn't actually the only exposure. `Task.Meta` and `Task.Config` also get flattened and diffed as raw strings via `fieldDiffs` (in `Task.Diff()` and `configDiff()`), with no JSON-detection at all — #23603 only special-cased `EmbeddedTmpl` by name (`diff.Name == “EmbeddedTmpl"`), it didn't touch `fieldDiff` itself. So a JSON-valued meta entry or driver config value would still hit the same unreadable-diff problem.

Given that, I don't think this closes cleanly as covered by #23603 — it actually points toward option (2) from my earlier comment: generalizing the JSON pretty-print into `fieldDiff` itself (keyed on "does this value look like JSON," not on field name), rather than leaving it as a name-gated special case. Want me to go ahead with that scope?

**Comment by tgross** [comment]
> `Task.Meta` and `Task.Config` also get flattened and diffed as raw strings... So a JSON-valued meta entry or driver config value would still hit the same unreadable-diff problem.

I'm not sure I understand the specific case you're trying to avoid here.  As far as I can tell it wouldn't be appropriate to try to detect JSON here because these values are not arbitrary large chunks of text but structured (albeit loosely by the plugin config schema or as `map[string]string`). Can you provide a jobspec that shows what you mean?

**Comment by vipulbhasin23** [comment]
Fair point on Meta/Config specifically — you're right those are structurally typed. But I found a concrete, verified case that shows the same underlying problem: `nomad/structs/diff_test.go`, in `TestTaskGroupDiff` (the `"TaskGroup Services edited”` case, ~line 3367), sets `ServiceCheck.Body` to `"{\"key\": \"value\”}"` — a JSON string, which is a legitimate real-world value for an HTTP health check body.

The expected diff output for that field (~line 3645) is a plain `FieldDiff` with the raw string in `Old`/`New` — no JSON-aware rendering, same as `EmbeddedTmpl` before #23603. So it looks like the problem isn't really about which container type holds the value (`map[string]string` vs. a plain string field) — it's about any field where the value itself can be a JSON blob, and `fieldDiff` doesn't inspect content for that.

Given `Check.Body` is a real, independent instance of this, does it change how you'd want to scope this — as a targeted `Check.Body` fix, a general content-based detection in `fieldDiff`, or something else?

**Comment by tgross** [comment]
So you don't have a use case, just something your LLM extracted from the tests. I'm marking this issue as fixed by #23603 once that's been merged. I'm not convinced it's ever going to be a good idea for Nomad to _guess_ at JSON contents of arbitrary string blobs, regardless of what Terraform did here.

**Comment by vipulbhasin23** [comment]
Fair enough — I'll leave this as-is for now. `Check.Body` was more of an illustration that the problem isn't unique to `EmbeddedTmpl`, not a real complaint.
